### DISCUSSION TITLE
Using Other Trial Model

### DISCUSSION THREAD

**Discussion by rm17tink** [optuna/optuna#4511] [Q&A] [answered]
### Don't use GitHub Issues to ask support questions.

Is it possible to get access or overwrite the saved best_booster with one of the other trials? 

`def callback(study, trial):
    global best_booster
    if study.best_trial == trial:
        best_booster = model   `

best Booster is trial 35 for instance

I know I can call the trial via:  
target_trial= 30
`target_trial = [trial for trial in study.trials if trial.number == trial_number][0]`

However, I want the entire(other trial 30) model object to use for scoring and saving and using it on my test group

Right now I can only use the best booster:

Scoring:
`best_booster.predict_proba(test[cols_for_model])`

Saving:

`joblib.dump(best_booster, 'model.pkl')`

Is there a convenient way to accomplish this with another trial model?

Thanks.

**Comment by nzw0301** [comment, marked as answer]
I might not fully understand the question, but how about saving the booster in the objective, for example `joblib.dump(booster, 'model_{}.pkl'.format(trial.number))` just before returning the objective value instead of using a callback,  and loading them after optimisation? I suppose that the best model can be loaded from the best trial number stored in `study` instance.

**Reply by rm17tink** [reply]
```
import lightgbm as lgb

import optuna


def objective(trial):
    param = {
        'objective': 'binary',
        'metric': 'binary_logloss',
        'verbosity': -1,
        'boosting_type': 'gbdt',
        'lambda_l1': trial.suggest_float('lambda_l1', 1e-8, 10.0, log=True),
        'lambda_l2': trial.suggest_float('lambda_l2', 1e-8, 10.0, log=True),
        'num_leaves': trial.suggest_int('num_leaves', 2, 256),
        'feature_fraction': trial.suggest_float('feature_fraction', 0.4, 1.0),
        'bagging_fraction': trial.suggest_float('bagging_fraction', 0.4, 1.0),
        'bagging_freq': trial.suggest_int('bagging_freq', 1, 7),
        'min_child_samples': trial.suggest_int('min_child_samples', 5, 100),
    }

    gbm = lgb.train(param, dtrain)

    **joblib.dump(gbm, 'model_{}.pkl'.format(trial.number))**

    return accuracy

study = optuna.create_study(direction='maximize')
study.optimize(objective, n_trials=100)
```

is this what you mean? Instead of saving only the best booster from the callback, i actually save all models in the objective then pull the model I assessed to be the best?

**Reply by nzw0301** [reply]
Yes.
